atef/config.py

```diff
@@ -157,14 +184,820 @@ def from_yaml(cls, filename: AnyPath) -> ConfigurationFile:
+        parent: Optional[PreparedGroup] = None,
+        cache: Optional[DataCache] = None,
+    ) -> PreparedPVConfiguration:
+        """
+        Ready a set of PV checks without requiring an existing PVConfiguration.
+
+        Parameters
+        ----------
+        by_pv : Dict[str, List[Comparison]]
+            PV name to comparison list.
+        shared : list of Comparison, optional
+            Comparisons to be run on *all* PVs in the `by_pv` dictionary.
+        parent : PreparedGroup, optional
+            The parent group.
+        cache : DataCache, optional
+            The data cache instance, if available.  If unspecified, a new data
+            cache will be instantiated.
+
+        Returns
+        -------
+        PreparedPVConfiguration
+
+        """
+        config = cls.from_config(
+            PVConfiguration(
+                by_pv=by_pv,
+                shared=shared or [],
+            ),
+            cache=cache,
+            parent=parent,
+        )
+        return cast(PreparedPVConfiguration, config)
+
+    @classmethod
+    def from_config(
+        cls,
+        config: PVConfiguration,
+        parent: Optional[PreparedGroup] = None,
+        cache: Optional[DataCache] = None,
+    ) -> PreparedPVConfiguration:
+        """
+        Prepare a PVConfiguration for running.
+
+        Parameters
+        ----------
+        config : PVConfiguration
+            The configuration settings.
+        parent : PreparedGroup, optional
+            The parent group.
+        cache : DataCache, optional
+            The data cache instance, if available.  If unspecified, a new data
+            cache will be instantiated.
+
+        Returns
+        -------
+        PreparedPVConfiguration
+        """
+        if not isinstance(config, PVConfiguration):
+            raise ValueError(f"Unexpected configuration type: {type(config).__name__}")
+
+        if cache is None:
+            cache = DataCache()
+
+        prepared_comparisons = []
+        prepare_failures = []
+        shared = config.shared or []
+
+        prepared = PreparedPVConfiguration(
+            config=config,
+            cache=cache,
+            parent=parent,
+            comparisons=prepared_comparisons,
+            prepare_failures=prepare_failures,
+        )
+
+        for pvname, comparisons in config.by_pv.items():
+            for comparison in comparisons + shared:
+                try:
+                    prepared_comparisons.append(
+                        PreparedSignalComparison.from_pvname(
+                            pvname=pvname,
+                            comparison=comparison,
+                            parent=prepared,
+                            cache=cache,
+                        )
+                    )
+                except Exception as ex:
+                    prepare_failures.append(ex)
+
+        return prepared
+
+
+@dataclass
+class PreparedToolConfiguration(PreparedConfiguration):
```

Comment:
It's a bit unfortunate that there has to be three extremely similar copies of these prepared configuration subclasses

Reply:
If you have suggestions as to how to restructure it, I'd be happy to revisit it

Reply:
I do not have any specific suggestions! I was just lamenting a bit

Reply:
No problem, I thought it was unfortunate while writing it, too...